core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1102,6 +1102,42 @@ public Builder setDefaultPartitionSpec(int specId) {
       return this;
     }
 
+    Builder removeUnusedSpecs(Iterable<PartitionSpec> specsToRemove) {
+      Set<Integer> specIdsToRemove = Sets.newHashSet();
+      for (PartitionSpec spec : specsToRemove) {
+        Preconditions.checkArgument(
+            spec.specId() != defaultSpecId, "Cannot remove default partition spec");
+        PartitionSpec toBeRemoved = specsById.remove(spec.specId());
+        Preconditions.checkArgument(
+            toBeRemoved == null || toBeRemoved.equals(spec),
+            "Cannot remove an unknown spec, spec id: %s",
+            spec.specId());
+        if (toBeRemoved != null) {
+          specIdsToRemove.add(spec.specId());
+        }
+      }
+      this.specs =
+          specs.stream()
+              .filter(s -> !specIdsToRemove.contains(s.specId()))
+              .collect(Collectors.toList());
+      changes.add(new MetadataUpdate.RemoveUnusedSpecs(specIdsToRemove));
+      return this;
+    }
+
+    Builder removeUnusedSpecsById(Iterable<Integer> specIds) {
+      Set<Integer> specIdsToRemove = Sets.newHashSet(specIds);
+      Preconditions.checkArgument(
+          !specIdsToRemove.contains(defaultSpecId), "Cannot remove default partition spec");
+      this.specs =
+          specs.stream()
+              .filter(s -> !specIdsToRemove.contains(s.specId()))
+              .collect(Collectors.toList());
+      specsById.clear();
+      specsById.putAll(indexSpecs(specs));
```

Comment:
I'm a little worried here that we are dropping the defaultSpec from the "specsById" map here

Reply:
Should this be "specs" or "this.specs"? I'm getting a little lost. Perhaps a different name is needed